crates/portal-logs-collector/src/server.rs

```diff
@@ -0,0 +1,137 @@
+    task_manager: TaskManager,
+    event_stream: Box<dyn Stream<Item = PortalLogsCollectorEvent> + Send + Unpin + 'static>,
+    collector_index: usize,
+    collector_group_size: usize,
+    _phantom: PhantomData<T>,
+}
+
+impl<T> Server<T>
+where
+    T: Storage + Send + Sync + 'static,
+{
+    pub fn new(
+        transport_handle: PortalLogsCollectorTransportHandle,
+        event_stream: impl Stream<Item = PortalLogsCollectorEvent> + Send + Unpin + 'static,
+        logs_collector: PortalLogsCollector<T>,
+        collector_index: usize,
+        collector_group_size: usize,
+    ) -> Self {
+        Self {
+            _transport_handle: transport_handle,
+            logs_collector,
+            registered_gateways: Default::default(),
+            task_manager: Default::default(),
+            event_stream: Box::new(event_stream),
+            collector_index,
+            collector_group_size,
+            _phantom: Default::default(),
+        }
+    }
+
+    pub async fn run(
+        mut self,
+        contract_client: Arc<dyn ContractClient>,
+        collection_interval: Duration,
+        portal_update_interval: Duration,
+        cancellation_token: CancellationToken,
+    ) -> anyhow::Result<()> {
+        log::info!("Starting logs collector server");
+
+        // Get registered gateways from chain
+        let gateways = contract_client
+            .active_gateways()
+            .await?
+            .into_iter()
+            .collect();
+        *self.registered_gateways.lock() = gateways;
+
+        self.spawn_portal_update_task(contract_client, portal_update_interval);
+
+        self.run_collecting_task(collection_interval, cancellation_token.child_token())
+            .await;
+
+        log::info!("Server shutting down");
+        self.task_manager.await_stop().await;
+        Ok(())
+    }
+
+    fn should_process(&self, peer_id: &PeerId) -> bool {
+        if let Some(byte) = peer_id.to_bytes().last() {
+            (*byte as usize) % self.collector_group_size == self.collector_index
+        } else {
+            false
+        }
+    }
+
+    async fn run_collecting_task(&mut self, interval: Duration, cancel_token: CancellationToken) {
+        let mut interval = tokio::time::interval(interval);
+        interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+        loop {
+            tokio::select! {
+                _ = interval.tick() => {
+                    let res = self.logs_collector.dump_buffer().await;
+                    let _ = res.inspect_err(|err| log::error!("Error while dumping records: {err:?}"));
+                },
+                Some(LogQuery { peer_id, log }) = self.event_stream.next() => {
+                    if !self.should_process(&peer_id) {
+                        continue
+                    }
+                    if self.registered_gateways.lock().contains(&peer_id) {
+                        log::error!("Got log from {peer_id:?}: {log:?}");
```

Comment:
It's also not an `error`. We can keep it at `info`, but at some point it will probably changed to `debug`